internal/descriptor/provider/provider.go

```diff
@@ -0,0 +1,96 @@
+package provider
+
+import (
+	"errors"
+
+	"github.com/open-component-model/ocm/pkg/contexts/ocm/compdesc"
+
+	"github.com/kyma-project/lifecycle-manager/api/v1beta2"
+	"github.com/kyma-project/lifecycle-manager/internal/descriptor/cache"
+)
+
+var (
+	errTypeAssert  = errors.New("failed to convert to v1beta2.Descriptor")
+	errDecode      = errors.New("failed to decode to descriptor target")
+	errTemplateNil = errors.New("module template is nil")
+)
+
+type CachedDescriptorProvider struct {
+	descriptorCache *cache.DescriptorCache
+}
+
+func NewCachedDescriptorProvider(descriptorCache *cache.DescriptorCache) *CachedDescriptorProvider {
+	if descriptorCache != nil {
+		return &CachedDescriptorProvider{
+			descriptorCache: descriptorCache,
+		}
+	}
+	return &CachedDescriptorProvider{
+		descriptorCache: cache.NewDescriptorCache(),
+	}
+}
+
+func (c *CachedDescriptorProvider) GetDescriptor(template *v1beta2.ModuleTemplate) (*v1beta2.Descriptor, error) {
+	if template.Spec.Descriptor.Object != nil {
+		desc, ok := template.Spec.Descriptor.Object.(*v1beta2.Descriptor)
+		if !ok {
+			return nil, errTypeAssert
+		}
+		return desc, nil
+	}
+	key := cache.GenerateDescriptorKey(template)
```

Comment:
Could be a one-liner (kind-of 😉). But it is subjective

```suggestion
	if descriptor := c.descriptorCache.Get(cache.GenerateDescriptorKey(template)); descriptor != nil {
		return descriptor, nil
	}
```